src/terminal/terminal.rs

```diff
@@ -234,39 +234,143 @@ where
+    /// a `Result` instead of nothing.
+    ///
+    /// Applications should call `try_draw` or [`draw`] in a loop to continuously render the
+    /// terminal. These methods are the main entry points for drawing to the terminal.
+    ///
+    /// [`draw`]: Terminal::draw
+    ///
+    /// This method will:
+    ///
+    /// - autoresize the terminal if necessary
+    /// - call the rendering parameter, passing it a [`Frame`] reference to render to
+    /// - flush the current internal state by copying the current buffer to the backend
+    /// - move the cursor to the last known position if it was set during the rendering closure
+    /// - return a [`CompletedFrame`] with the current buffer and the area of the terminal
+    ///
+    /// The render function passed to `try_draw` can return any a [`Result`] with an error type that
+    /// can be converted into an [`std::io::Error`] using the [`Into`] trait. This makes it possible
+    /// to use the `?` operator to propagate errors that occur during rendering. If the render
+    /// function returns an error, the error will be returned from `try_draw` as an
+    /// [`std::io::Error`] and the terminal will not be updated.
+    ///
+    /// The [`CompletedFrame`] returned by this method can be useful for debugging or testing
+    /// purposes, but it is often not used in regular applicationss.
+    ///
+    /// The render function should fully render the entire frame when called, including areas that
+    /// are unchanged from the previous frame. This is because each frame is compared to the
+    /// previous frame to determine what has changed, and only the changes are written to the
+    /// terminal. If the render function does not fully render the frame, the terminal will not be
+    /// in a consistent state.
+    ///
+    /// # Examples
+    ///
+    /// ```should_panic
+    /// # let backend = ratatui::backend::TestBackend::new(10, 10);
+    /// # let mut terminal = ratatui::Terminal::new(backend)?;
+    /// use std::io;
+    ///
+    /// use ratatui::widgets::Paragraph;
+    ///
+    /// // with a closure
+    /// terminal.try_draw(|frame| {
+    ///     let value: u8 = "not a number".parse().map_err(io::Error::other)?;
+    ///     let area = frame.size();
+    ///     frame.render_widget(Paragraph::new("Hello World!"), area);
+    ///     frame.set_cursor(0, 0);
+    ///     io::Result::Ok(())
+    /// })?;
+    ///
+    /// // or with a function
+    /// terminal.try_draw(render)?;
+    ///
+    /// fn render(frame: &mut ratatui::Frame) -> io::Result<()> {
+    ///     let value: u8 = "not a number".parse().map_err(io::Error::other)?;
+    ///     frame.render_widget(Paragraph::new("Hello World!"), frame.size());
+    ///     Ok(())
+    /// }
+    /// # io::Result::Ok(())
+    /// ```
+    pub fn try_draw<F, E>(&mut self, render: F) -> io::Result<CompletedFrame>
+    where
+        F: FnOnce(&mut Frame) -> Result<(), E>,
+        E: Into<io::Error>,
```

Comment:
Is `io::Error` the right error here? I'm looking at the variants in [`ErrorKind`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html) and it doesn't feel like a lot of these variants will be used. Is there a CustomError type we should introduce for this? Can we make it `Box<dyn Err>`?

Reply:
Alternatively, we can introduce this as experimental until we explore this more?

Reply:
All of the crossterm errors are io Errors. There is an io::Error::Other that can be used to represent anything that doesn't fit. try_draw and draw shouldn't diverge in their error types, as the complexity is unwarranted and would be unpredictable / weird.

I've thought about someday introducing a proper error type for ratatui, but this is orthogonal to this. Because most apps will use draw (due to inertia), we are no worse off by adding try_draw with io::Error (it locks us into a error type that we're already locked into).

Put another way, draw has to return io::Error, and the things which try_draw call return that too. There's nothing we'd gain by using some other error. Note - the E: Into<io::Error> means that all we're saying is that the error that the callback returns must be convertible into the errors that we need as input and output of the various parts.